### ISSUE TITLE
Improving Language Understanding by Generative Pre-Training

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#14]
# Key Idea

- **Unsupervised Pre-Training**
- **Supervised Fine-Tuning**
- **Auxiliary Training Objectives**

# Previous Approaches

### Semi-Supervised Learning

- Word Embedding from Unsupervised Corpora + Supervised Downstream Task
- Recently, higher level semantics, such as phrase / sentence level representations are being studied.

### Unsupervised Pre-Training

- Used in diverse domains, such as Image Classification, NLP, etc.
- So far, unsupervised pre-training was not shown effective due to their use of LSTM, which is incapable of learning long-range context.

### Auxiliary Training Objectives

- POS tagging, chunking, named-entity recognition etc to improve semantic role labeling.

# Formulation

### Unsupervised Pre-Training

- 12 Transformer Decoder Blocks

![image](https://user-images.githubusercontent.com/46308407/104117752-0bd24a80-5367-11eb-9ca6-2d502e55f694.png)

![image](https://user-images.githubusercontent.com/46308407/104117756-14c31c00-5367-11eb-8b93-2f3bf0b78df3.png)

### Supervised Fine-Tuning

- Linear Output Layer를 기존의 Decoder Blocks 위에 추가

![image](https://user-images.githubusercontent.com/46308407/104117763-29071900-5367-11eb-8b8b-058e3e88790f.png)

![image](https://user-images.githubusercontent.com/46308407/104117772-32908100-5367-11eb-8af8-72d846b63a0f.png)

- Auxiliary objective helped fine-tuning by
  1. Improving generalization quality
  2. Boosting convergence

![image](https://user-images.githubusercontent.com/46308407/104117783-49cf6e80-5367-11eb-9b80-5320b04f1f4f.png)

### Task-specific Input Transformation

![image](https://user-images.githubusercontent.com/46308407/104117800-64094c80-5367-11eb-9bca-a9c558ace75a.png)

- **"Traversal Style Approach"**
    - Do not build separate architectures for downstream tasks
    - Convert structured inputs into an ordered sequence that base model can take in
    - Use $ sign as delimiter in some tasks.
- Textual Entailment
    - Concat: Premise$Hypo
- Similarity
    - Make sequences for all ordering permutation, get representations for each independently, add them element-wise and push through linear layer.
- Question Answering
    - Concat text-question-answer $[z;q;$$$;a_1]$, ..., $[z;q;$$$;a_k]$ and pass softmax. (k지선다)

# Specifics

- **Unsupervised Pre-Training**
- BooksCorpus Dataset, which contains 7,000 books of different genres. : "***Long Stretches of Contiguous Text"***
- 12 Transformer Decoders with 12 attention heads
- Adam Optimizer with max lr 2.5e-4
- Because of Layer Norm, simple initialization of $N(0, 0.02)$ was sufficient
- BPE for word-level encoding / GELU as activation

- **Supervised Fine-Tuning**
- lr 6.25e-5, dropout 0.1, 3 epochs

# Experiment Result

![image](https://user-images.githubusercontent.com/46308407/104117804-74212c00-5367-11eb-98aa-3b146d193a28.png)

**Implications**

- (left) Each transformer layer provides useful information for downstream task
- (right) The model learns effective representations only with Generative Pre-Training

![image](https://user-images.githubusercontent.com/46308407/104117807-813e1b00-5367-11eb-8d4d-82f7d0d6f040.png)

![image](https://user-images.githubusercontent.com/46308407/104117811-88fdbf80-5367-11eb-9995-d2f29f862d5b.png)

![image](https://user-images.githubusercontent.com/46308407/104117814-90bd6400-5367-11eb-9fb4-0649a9f429bf.png)
